RenderTable::BottomSection() should return the last rendered table section

karlcow opened the pull request:
#### 1d539b1ae929bb6062b9501cc61c97d57c0108b1
<pre>
RenderTable::BottomSection() should return the last rendered table section
<a href="https://bugs.webkit.org/show_bug.cgi?id=257575">https://bugs.webkit.org/show_bug.cgi?id=257575</a>
<a href="https://rdar.apple.com/110430887">rdar://110430887</a>

Reviewed by Alan Baradlay.

BottomSection() was not necessary returning the last rendered section.
In return it had an impact on drawing the border of the table element
when the borders are collapsed.
See <a href="https://searchfox.org/wubkat/rev/47b66cb0dc7c2146fa7576555c720080708667f3/Source/WebCore/rendering/RenderTable.cpp#1358-1378">https://searchfox.org/wubkat/rev/47b66cb0dc7c2146fa7576555c720080708667f3/Source/WebCore/rendering/RenderTable.cpp#1358-1378</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border.html: Added.
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::bottomSection const):

Canonical link: <a href="https://commits.webkit.org/289434@main">https://commits.webkit.org/289434@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2dbb451b1b890005ebe61b561bc8cd777529e403

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86968 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6477 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91826 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37705 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14545 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67224 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24991 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89970 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5155 "Failed to checkout and rebase branch from PR 39589") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78717 "Failed to checkout and rebase branch from PR 39589") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47546 "Failed to checkout and rebase branch from PR 39589") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4935 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33089 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36820 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75430 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33968 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93712 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14128 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10273 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76029 "Failed to checkout and rebase branch from PR 39589") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74570 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75226 "Failed to checkout and rebase branch from PR 39589") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19552 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17977 "Failed to checkout and rebase branch from PR 39589") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7007 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13539 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14147 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19438 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13892 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17334 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15674 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1474,7 +1474,11 @@ RenderTableSection* RenderTable::bottomSection() const
+            continue;
         if (auto* tableSection = dynamicDowncast<RenderTableSection>(*child))
             return tableSection;
     }
```
This looks good, though I would move the entire function next to `RenderTable::topSection` and mirror its logic something along the lines of.

```
RenderTableSection* RenderTable::bottomSection() const
{
    recalcSectionsIfNeeded();
    if (m_foot)
        return m_foot.get();
    for (CheckedPtr child = lastChild(); child; child = child->previousSibling()) {
        if (child.get() == m_head.get())
            continue;
        if (auto* tableSection = dynamicDowncast<RenderTableSection>(*child))
            return tableSection;
    }
    return m_head.get();
}

```

so it's like, return footer first, body second and head last.

karlcow replied:
moved and changed!
but github issues for now to send the PR.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border-ref.html:
```diff
@@ -0,0 +1,42 @@
+<title>CSS Test: Table border is correct even with out of order tbody, thead, tfoot</title>
+<link rel="author" title="Karl Dubost" href="mailto:karlcow@apple.com">
+<link rel="help" href="https://html.spec.whatwg.org/multipage/tables.html#forming-a-table">
+<link rel="match" href="out-of-order-elements-collapsed-border-ref.html">
```
These are not useful in a -ref.html file :) (same comment about -expected.html)

karlcow replied:
nice catch! Thanks.

karlcow replied:
changed!

karlcow committed:
RenderTable::BottomSection() should return the last rendered table section
https://bugs.webkit.org/show_bug.cgi?id=257575
rdar://110430887

Reviewed by Alan Baradlay.

BottomSection() was not necessary returning the last rendered section.
In return it had an impact on drawing the border of the table element
when the borders are collapsed.
See https://searchfox.org/wubkat/rev/47b66cb0dc7c2146fa7576555c720080708667f3/Source/WebCore/rendering/RenderTable.cpp#1358-1378

* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/out-of-order-elements-collapsed-border.html: Added.
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::bottomSection const):

Canonical link: https://commits.webkit.org/289434@main
